### DISCUSSION TITLE
How should we handle JsonSerializer.DeserializeAsync nullable return type?

### DISCUSSION THREAD

**Discussion by 0xced** [dotnet/runtime#60195] [Q&A] [answered]
The [`System.Text.Json.JsonSerializer.DeserializeAsync<TValue>`](https://github.com/dotnet/runtime/blob/v6.0.0/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L43) method returns a nullable type (`TValue?`).

The [documentation](https://docs.microsoft.com/en-us/dotnet/api/System.Text.Json.JsonSerializer.DeserializeAsync#System_Text_Json_JsonSerializer_DeserializeAsync__1_System_IO_Stream_System_Text_Json_JsonSerializerOptions_System_Threading_CancellationToken_) says
> **Returns**
> ValueTask&lt;TValue&gt;
> A `TValue` representation of the JSON value.

But the actual return type is `ValueTask<TValue?>` and the documentation *forgets* to mention that. So, what does that mean? How can a result be `null` instead of an exception being thrown during deserialization?

I'm currently throwing a `JsonException` but I think it's pretty bad because the exception message is rather useless.
```csharp
var mediaInfo = await JsonSerializer.DeserializeAsync<MediaInfo>(stream, cancellationToken);
if (mediaInfo is null)
{
    throw new JsonException("Failed to deserialize the JSON stream as MediaInfo.");
}
```

How are we supposed to handle `JsonSerializer.DeserializeAsync` returning `null`?

**Comment by FiniteReality** [comment]
I honestly think it would be *way* more practical if `Deserialize` and co. only returned a nullable value if the input was nullable. Pretty much all of my `Deserialize`/`DeserializeAsync` calls have the same guard on them.

**Comment by huoyaoyuan** [comment, marked as answer]
> So, what does that mean? How can a result be null instead of an exception being thrown during deserialization?

If you pass a root literal null (the json text is `null` without quotation), it should be deserialized as null.

Generally, you should always verify bad input. Sometimes it's correctly formed, but the content of fields are bad. Nulls can be handled in the verification, together with null fields.

**Reply by 0xced** [reply]
I did not think about a root literal `null`! It makes sense but it would be worth documenting explicitly in the `System.Text.Json.JsonSerializer.DeserializeAsync<TValue>` documentation.
